packages/virtual-list/src/vaadin-virtual-list-selection-mixin.js

```diff
@@ -0,0 +1,435 @@
+  class SelectionMixin extends superClass {
+    static get properties() {
+      return {
+        /**
+         * Selection mode for the virtual list. Available modes are: `none`, `single` and `multi`.
+         * @attr {string} selection-mode
+         */
+        selectionMode: {
+          type: String,
+          observer: '__selectionModeChanged',
+          value: 'none',
+        },
+
+        /**
+         * Path to an item sub-property that identifies the item.
+         * @attr {string} item-id-path
+         */
+        itemIdPath: {
+          type: String,
+          value: null,
+          sync: true,
+        },
+
+        /**
+         * An array that contains the selected items.
+         * @type {!Array<!VirtualListItem>}
+         */
+        selectedItems: {
+          type: Array,
+          notify: true,
+          value: () => [],
+          sync: true,
+        },
+
+        /**
+         * A function that generates accessible names for virtual list items.
+         */
+        itemAccessibleNameGenerator: {
+          type: Function,
+          sync: true,
+        },
+
+        /**
+         * Set of selected item ids
+         * @private
+         */
+        __selectedKeys: {
+          type: Object,
+          computed: '__computeSelectedKeys(itemIdPath, selectedItems)',
+        },
+
+        /**
+         * The index of the focused item.
+         * @private
+         */
+        __focusIndex: {
+          type: Number,
+          value: 0,
+          sync: true,
+        },
+      };
+    }
+
+    static get observers() {
+      return ['__selectionChanged(itemIdPath, selectedItems, itemAccessibleNameGenerator)'];
+    }
+
+    constructor() {
+      super();
+
+      this.addEventListener('keydown', (e) => this.__onKeyDown(e));
+      this.addEventListener('click', (e) => this.__onClick(e));
+      this.addEventListener('focusin', (e) => this.__onFocusIn(e));
+      this.addEventListener('focusout', (e) => this.__onFocusOut(e));
+    }
+
+    ready() {
+      super.ready();
+
+      this._createPropertyObserver('items', '__selectionItemsUpdated');
+    }
+
+    /** @private */
+    __isSelectable() {
+      return this.selectionMode !== 'none';
+    }
+
+    /**
+     * @private
+     * @override
+     */
+    __updateElement(el, index) {
+      const item = this.items[index];
+      el.__item = item;
+      el.__index = index;
+
+      el.toggleAttribute('selected', this.__isSelected(item));
+      const isFocusable = this.__isNavigating() && this.__focusIndex === index;
+      el.tabIndex = isFocusable ? 0 : -1;
+      el.toggleAttribute(
+        'focused',
+        this.__isSelectable() && this.__focusIndex === index && el.contains(this.__getActiveElement()),
+      );
+
+      el.role = this.__isSelectable() ? 'option' : 'listitem';
+      el.ariaSelected = this.__isSelectable() ? String(this.__isSelected(item)) : null;
+      el.ariaSetSize = String(this.items.length);
```

Comment:
It seems like SelectionMixin currently takes responsibilities that are beyond its intended scope, e.g. setting `aria-set-size` or `aria-pos-in-set`. I'm suggesting that we create a new mixin, something like `VirtualizerMixin`, which would handle the default `__updateElement` logic. Then, SelectionMixin would override this method to add its selection-specific properties only when selection is enabled.

Reply:
Would it make sense to assign `aria-set-size` and `aria-pos-in-set ` in `virtual-list-mixin`'s `__updateElement`? I think the `selection-mixin`'s `__updateElement` still needs to run even if `selectionMode` isn't set to cover the cases where:
- `selectionMode` is set from `multi` to `none`
- `selectedItems` is programmatically populated

Reply:
Extracted a11y-related changes (on non-selectable mode) as a [separate PR](https://github.com/vaadin/web-components/pull/8328)